framework/lib/hooks/use-outside-rect-click/index.ts

```diff
@@ -0,0 +1,31 @@
+import { isNil } from 'ramda'
+import { MutableRefObject, useEffect } from 'react'
+
+const isBetween = (point: number, min: number, max: number) => (point >= min && point <= max)
+
+export const useOutsideRectClick =
+(ref: MutableRefObject<HTMLElement | null>, callback: () => void) =>
+  useEffect(() => {
+    const handleMouseClick = (event: MouseEvent) => {
+      const { clientX, clientY } = event
+      if (isNil(ref?.current)) return
+
+      const {
+        top,
+        right,
+        left,
+        bottom,
+      } = ref.current.getBoundingClientRect()
+
+      const isInBound = isBetween(clientX, left, right) && isBetween(clientY, top, bottom)
+
+      if (isInBound) return
+      callback()
+    }
+
+    document.addEventListener('click', handleMouseClick)
```

Comment:
I think you need both click and touchstart here and when removing the listener for it to work on for example iPads `document.addEventListener('click touchstart', handleMouseClick)`

Reply:
For reference see Chakra-UIs implementation of useOutsideClick: https://github.com/chakra-ui/chakra-ui/blob/main/packages/hooks/use-outside-click/src/index.ts